v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/jdbc/dialectadapter/postgresql/PostgreSQLDialectAdapter.java

```diff
@@ -0,0 +1,495 @@
+            final long numericScale = resultSet.getLong("numeric_scale");
+            boolean typeHasScale = !resultSet.wasNull();
+            if (typeHasMaximumCharacterLength) {
+              sourceColumnType =
+                  new SourceColumnType(columnType, new Long[] {characterMaximumLength}, null);
+            } else if (typeHasPrecision && typeHasScale) {
+              sourceColumnType =
+                  new SourceColumnType(
+                      columnType, new Long[] {numericPrecision, numericScale}, null);
+            } else if (typeHasPrecision) {
+              sourceColumnType =
+                  new SourceColumnType(columnType, new Long[] {numericPrecision}, null);
+            } else {
+              sourceColumnType = new SourceColumnType(columnType, new Long[] {}, null);
+            }
+            schema.put(columnName, sourceColumnType);
+          }
+          tableSchemaBuilder.put(table, schema.build());
+        }
+      }
+    } catch (SQLTransientConnectionException e) {
+      logger.warn(
+          "Transient connection error while discovering table schema for datasource={} db={} tables={}",
+          dataSource,
+          sourceSchemaReference,
+          tables,
+          e);
+      schemaDiscoveryErrors.inc();
+      throw new RetriableSchemaDiscoveryException(e);
+    } catch (SQLNonTransientConnectionException e) {
+      logger.error(
+          "Non Transient connection error while discovering table schema for datasource={}, db={} tables={}",
+          dataSource,
+          sourceSchemaReference,
+          tables,
+          e);
+      schemaDiscoveryErrors.inc();
+      throw new SchemaDiscoveryException(e);
+    } catch (SQLException e) {
+      logger.error(
+          "Sql exception while discovering table schema for datasource={} db={} tables={}",
+          dataSource,
+          sourceSchemaReference,
+          tables,
+          e);
+      schemaDiscoveryErrors.inc();
+      throw new SchemaDiscoveryException(e);
+    } catch (SchemaDiscoveryException e) {
+      // Already logged.
+      schemaDiscoveryErrors.inc();
+      throw e;
+    }
+    ImmutableMap<String, ImmutableMap<String, SourceColumnType>> tableSchema =
+        tableSchemaBuilder.build();
+    logger.info(
+        "Discovered tale schema for Datasource: {}, SourceSchemaReference: {}, tables: {}, schema: {}",
+        dataSource,
+        sourceSchemaReference,
+        tables,
+        tableSchema);
+
+    return tableSchema;
+  }
+
+  /**
+   * Discover the indexes of tables to migrate.
+   *
+   * @param dataSource Provider for JDBC connection.
+   * @param sourceSchemaReference Source database name and (optionally namespace)
+   * @param tables Tables to migrate.
+   * @return The discovered indexes.
+   * @throws SchemaDiscoveryException - Fatal exception during Schema Discovery.
+   * @throws RetriableSchemaDiscoveryException - Retriable exception during Schema Discovery.
```

Comment:
[super-nit], please feel free to ignore. For ease of review / future-reading in case of some of the MYSQL discovery queries, especially around index discover (which involves a join), we have added links to db-fiddle .

In case it's not a lot of trouble or if you have them ready from your experiments, feel free to add them for PG too.

Reply:
That is very interesting, added and discovered that one query did not work with PG < 11. Added todo for the latter.